server/repo-op.c

```diff
@@ -1706,30 +1709,58 @@ put_dirent_and_commit (SeafRepo *repo,
                        const char *path,
                        SeafDirent *dent,
                        int replace,
+                       int file_num,
                        const char *user,
                        GError **error)
 {
     SeafCommit *head_commit = NULL;
+    SeafDirent **dst_dents = NULL;
     char *root_id = NULL;
     char buf[SEAF_PATH_MAX];
-    int ret = 0;
+    int ret = 0, i = 0;
 
     GET_COMMIT_OR_FAIL(head_commit, repo->id, repo->version, repo->head->commit_id);
+    
+    /* If file_num > 0, dent is a second-rank-pointer which point to all the dirents */
+    if (file_num > 0) {
+        dst_dents = (SeafDirent **)dent;
+        root_id = head_commit->root_id;
+        for (i = 0; i < file_num; i++) {
+            char *tmp_root_id = root_id;
+            root_id = do_post_file_replace (repo,
```

Comment:
You should use do_post_multi_files() here. Calling do_post_file_replace() in a loop will  update objects a lot of times.